### PR TITLE
[ENH] Dask backend aggregations

### PR DESCRIPTION
**Pull request by gerrymanoim**
This PR does a few things:

1. Enables aggregations in the backend and adds missing test coverage for grouped/masked reductions
2. Changes dask `conftest.py`  objects to return `DataFrame` objects with `npartitions=2` to make sure we are exercising multi partition behavior. 
3. Fixes bugs uncovered by the change in `(2)`. 
4. Refactors backend testing fixtures so we don't need `if backend.name() == "dask"` calls in `ibis/backends/tests/test_aggregation.py`. 

- [ ]  On approval/merge update https://github.com/ibis-project/ibis/issues/2553

### CODE REVIEW

File: ibis/backends/dask/execution/util.py [hunk 1]
```diff
@@ -37,3 +39,24 @@ def make_selected_obj(gs: SeriesGroupBy):
         return gs.obj.set_index(gs.index, drop=False)[
             gs._meta._selected_obj.name
         ]
+
+
+def maybe_wrap_scalar(result: Any, expr: ir.Expr) -> Any:
```
Comment (jreback) [inline ibis/backends/dask/execution/util.py:44]: can we call coerce_to_output then

File: ibis/backends/tests/test_aggregation.py [hunk 2]
```diff
@@ -56,27 +56,25 @@ def mean_udf(s):
...
-    ['dask']
-)  # TODO - aggregations - #2553 (and pd.concat)
 @pytest.mark.xfail_unsupported
 def test_aggregate(
     backend, alltypes, df, result_fn, expected_fn, expected_col
 ):
     expr = alltypes.aggregate(tmp=result_fn)
     result = expr.execute()
 
+    if backend.name() == 'dask':
```
Comment (jreback) [inline ibis/backends/tests/test_aggregation.py:66]: maybe instead skip for dask and add a dask specific test? not sure what the pattern is for others like this

Comment (jreback) [reply ibis/backends/tests/test_aggregation.py:66]: or can have the backend assertion do t he compute?

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/backends/tests/test_aggregation.py (hunk 2 above)
Comment (gerrymanoim) [reply ibis/backends/tests/test_aggregation.py:66]: Yeah - the problem here specifically was the use of a lambda in `agg` which pandas knew how to handle but dask did not. The bigger problem I guess is using pandas inside the test itself.

File: ibis/backends/dask/execution/util.py (hunk 1 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:44]: I'm not convinced yet that we'll need it (or that dask udfs are going to operate like pandas udfs) so I picked I name that reflected the existing functionality.

File: ibis/backends/dask/execution/util.py [hunk 3]
```diff
@@ -37,3 +39,24 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+    A partial implementation of `coerce_to_output` in the pandas backend.
+
+    Currently only wraps scalars, but will change when udfs are added to the
+    dask backend.
+    """
+    result_name = expr.get_name()
+    if isinstance(result, dd.core.Scalar) and isinstance(
+        expr.op(), ops.Reduction
+    ):
+        # I don't believe we can avoid computing if we want to wrap this in a
```
Comment (jreback) [inline ibis/backends/dask/execution/util.py:55]: add aconsistent comment so can revist later, e.g. TODO

File: ibis/backends/tests/test_aggregation.py (hunk 2 above)
Comment (jreback) [reply ibis/backends/tests/test_aggregation.py:66]: ok can we instead have a property / function on the backend testing class that that allows us to do this automatically? (e.g. for all others this is a no-op).

**Review by jreback — CHANGES_REQUESTED** [review]

File: ibis/backends/tests/test_aggregation.py (hunk 2 above)
Comment (gerrymanoim) [reply ibis/backends/tests/test_aggregation.py]: I've replaced all of the checks except for one with a transparent `pandas_df` fixture. 

I left the one stopping us from attempting to sort result in dask. I'll remove that when I tackle sorting.

**Comment by jreback** [comment]
what was issue on 3.7 dask builds?

**Comment by gerrymanoim** [comment]
> what was issue on 3.7 dask builds?

It was actually a difference in dask versions. When you join a dataframe and a series, newer versions of dask will call `to_frame` on the joined series, but that wasn't the case in 2.22 (our min version). I just called it myself.

File: ibis/backends/dask/tests/conftest.py [hunk 4]
```diff
@@ -54,15 +54,30 @@ def assert_series_equal(
         kwargs.setdefault('check_dtype', cls.check_dtype)
         kwargs.setdefault('check_names', cls.check_names)
         # we sometimes use pandas to build the "expected" case in tests
-        right = right.compute() if isinstance(right, dd.Series) else right
-        tm.assert_series_equal(left.compute(), right, *args, **kwargs)
+        incoming_types = tuple(map(type, (left, right)))
+        if incoming_types == (dd.Series, pd.Series):
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/tests/conftest.py:58]: Because I only want to do a `reset_index` behavior if the left one was created in dask and has more than one partition, while the right one was created in pandas. In all other circumstances we can just `compute()` any dask output.

File: ibis/backends/dask/execution/util.py [hunk 5]
```diff
@@ -37,3 +39,43 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+    dfs: List[Union[dd.Series, dd.DataFrame]], lens: List[int]
+) -> dd.DataFrame:
+    """
+    When axis=1 and divisions are unknown, Dask DataFrame.concat can only
+    operate on objects with equal lengths, otherwise it will raise a
+    ValueError in `concat_and_check`.
+
+    See https://github.com/dask/dask/blob/2c2e837674895cafdb0612be81250ef2657d947e/dask/dataframe/multi.py#L907 # noqa
+
+    Note - this is likely to be quite slow, but this should be hit rarely in
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:74]: I'll add a note. This only occurs when you have  multiple grouped aggregates, some with filters and some without (thus you end up assembling disparate Series of different lens).

File: ibis/backends/dask/execution/aggregations.py [hunk 6]
```diff
@@ -72,10 +79,17 @@ def execute_aggregation_dataframe(
     pieces = []
     for metric in op.metrics:
         piece = execute(metric, scope=scope, timecontext=timecontext, **kwargs)
-        piece.name = metric.get_name()
+        piece = maybe_wrap_scalar(piece, metric)
         pieces.append(piece)
 
-    result = dd.concat(pieces, axis=1)
+    # We must perform this check here otherwise dask will throw a ValueError
+    # on `concat_and_check`. See docstring on `util.concat_via_join` for
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/execution/aggregations.py:86]: Sure - I can change this to a `utils.concat` and perform the dispatch inside there.

File: ibis/backends/dask/execution/util.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:77]: > this is faster if you do a binary reduction of small pieces and walk up the tree (but of course that's just a todo)

Do you mean do the regular concat on equal length groups and then combine those reduced groups with a join?

[the remark this answers is not in the document]
Comment (jreback) [reply ibis/backends/dask/execution/util.py:77]: so i think (again depending on the perf hit / effort  / how much this actually matters), you could try 2 strategies here:

- join, but with a binary reduction (you don't keep joing to the same frame, rather you can do pairs, then do pairs of those, etc)
- use concat, but first reindex each piece so this works

[the remark this answers is not in the document]
Comment (jreback) [reply ibis/backends/dask/execution/util.py:77]: but this can certainly be a TODO

**Review by jreback — APPROVED** [review]
lgtm

**Comment by jreback** [comment]
cc @icexelloss for a look

**Comment by jreback** [comment]
> > what was issue on 3.7 dask builds?
> 
> It was actually a difference in dask versions. When you join a dataframe and a series, newer versions of dask will call `to_frame` on the joined series, but that wasn't the case in 2.22 (our min version). I just called it myself. See [dask/dask#6809](https://github.com/dask/dask/pull/6809)

so our min version is the same right?

**Comment by gerrymanoim** [comment]
> > > what was issue on 3.7 dask builds?
> > 
> > 
> > It was actually a difference in dask versions. When you join a dataframe and a series, newer versions of dask will call `to_frame` on the joined series, but that wasn't the case in 2.22 (our min version). I just called it myself. See [dask/dask#6809](https://github.com/dask/dask/pull/6809)
> 
> so our min version is the same right?

Yep - I haven't touched the min version, just called `to_frame` myself.

File: ibis/backends/dask/execution/aggregations.py [hunk 7]
```diff
@@ -107,37 +117,32 @@ def execute_aggregation_dataframe(
...
-#     if isinstance(aggcontext, (agg_ctx.Summarize, agg_ctx.Transform)):
-#         result = aggcontext.agg(data, type(op).__name__.lower())
-#     else:
-#         result = aggcontext.agg(
-#             data, lambda data: getattr(data, type(op).__name__.lower())()
-#         )
-#     return result
+@execute_node.register((ops.Any, ops.All), (dd.Series, ddgb.SeriesGroupBy))
+def execute_any_all_series(op, data, aggcontext=None, **kwargs):
+    if isinstance(aggcontext, (agg_ctx.Summarize, agg_ctx.Transform)):
```
Comment (icexelloss) [inline ibis/backends/dask/execution/aggregations.py:122]: What are the two branches trying to handle differently?

File: ibis/backends/dask/execution/reductions.py [hunk 8]
```diff
@@ -73,37 +74,31 @@ def dask_execute_node_least_list(op, value, **kwargs):
...
-# TODO - aggregations - #2553
 @execute_node.register(ops.Reduction, ddgb.SeriesGroupBy, ddgb.SeriesGroupBy)
 def execute_reduction_series_gb_mask(
     op, data, mask, aggcontext=None, **kwargs
 ):
-    method = operator.methodcaller(type(op).__name__.lower())
-    return aggcontext.agg(
-        data, functools.partial(_filtered_reduction, mask.obj, method)
-    )
+    grouped_and_filtered_data = _filtered_reduction(data, mask)
```
Comment (icexelloss) [inline ibis/backends/dask/execution/reductions.py:92]: Why do we need to filter here for reduction?

File: ibis/backends/dask/execution/generic.py [hunk 9]
```diff
@@ -145,17 +145,19 @@
...
     Note: we cannot use the pandas version because Dask does not support .iloc
+    See https://docs.dask.org/en/latest/dataframe-indexing.html. .loc will
+    only work if our index lines up with the label.
     """
+    data = data[mask] if mask is not None else data
     if op.how == 'first':
         index = 0
     elif op.how == 'last':
-        index = -1
+        index = len(data) - 1  # TODO - computation
```
Comment (icexelloss) [inline ibis/backends/dask/execution/generic.py:155]: How is groupby handled here?

Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: Also, can you create an Jira to track this issue?

File: ibis/backends/dask/execution/reductions.py [hunk 10]
```diff
@@ -73,37 +74,31 @@ def dask_execute_node_least_list(op, value, **kwargs):
...
 def execute_reduction_series_groupby(
     op, data, mask, aggcontext=None, **kwargs
 ):
     return aggcontext.agg(data, type(op).__name__.lower())
 
 
-def _filtered_reduction(mask, method, data):
-    return method(data[mask[data.index]])
+def _filtered_reduction(data, mask):
+    return make_selected_obj(data)[mask.obj].groupby(data.index)
```
Comment (icexelloss) [inline ibis/backends/dask/execution/reductions.py:85]: The original pandas version doesn't have groupby, so I am a bit confused what this method is supposed to do.

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: The documentation says "Dask DataFrame does not track the length of partitions, making positional indexing with .iloc inefficient for selecting rows"

Does that mean here we trigger two pass of computation, one pass to compute the length, the second pass to index the row given the length? (I'd imagine this would be a full scan given what the documentation is describing)

File: ibis/backends/dask/execution/util.py [hunk 11]
```diff
@@ -37,3 +39,55 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+    """
+    result_name = expr.get_name()
+    if isinstance(result, dd.core.Scalar) and isinstance(
+        expr.op(), ops.Reduction
+    ):
+        # TODO - computation
+        return dd.from_pandas(
+            pd.Series(result.compute(), name=result_name), npartitions=1
+        )
+        # return dd.Series(result, name=result_name)
```
Comment (icexelloss) [inline ibis/backends/dask/execution/util.py:59]: Remove this?

File: ibis/backends/dask/execution/util.py [hunk 12]
```diff
@@ -37,3 +39,55 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+    A partial implementation of `coerce_to_output` in the pandas backend.
+
+    Currently only wraps scalars, but will change when udfs are added to the
+    dask backend.
+    """
+    result_name = expr.get_name()
+    if isinstance(result, dd.core.Scalar) and isinstance(
+        expr.op(), ops.Reduction
+    ):
+        # TODO - computation
```
Comment (icexelloss) [inline ibis/backends/dask/execution/util.py:55]: Please create a Jira to track this

File: ibis/backends/dask/execution/util.py [hunk 13]
```diff
@@ -37,3 +39,55 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+            pd.Series(result.compute(), name=result_name), npartitions=1
+        )
+        # return dd.Series(result, name=result_name)
+    else:
+        return result.rename(result_name)
+
+
+def safe_concat(dfs: List[Union[dd.Series, dd.DataFrame]]) -> dd.DataFrame:
+    """
+    Concat a list of `dd.Series` or `dd.DataFrame` objects into one DataFrame
```
Comment (icexelloss) [inline ibis/backends/dask/execution/util.py:66]: Are the Series and DataFrame aligned w.r.t partitiong?

File: ibis/backends/dask/execution/aggregations.py (hunk 7 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/aggregations.py:122]: https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/aggcontext.py as the different types of aggregation contexts that the pandas backend supports, which the dask backend currently uses. 

Right now the dask backend has no windowing, so the second branch isn't hit, but I left it in there to make starting on the windowing piece easier. I suspect we're going to have to treat this the same way.

File: ibis/backends/dask/execution/util.py [hunk 14]
```diff
@@ -37,3 +39,55 @@ def make_selected_obj(gs: SeriesGroupBy):
...
+    When axis=1 and divisions are unknown, Dask `DataFrame.concat` can only
+    operate on objects with equal lengths, otherwise it will raise a
+    ValueError in `concat_and_check`.
+
+    See https://github.com/dask/dask/blob/2c2e837674895cafdb0612be81250ef2657d947e/dask/dataframe/multi.py#L907 # noqa
+
+    Note - this is likely to be quite slow, but this should be hit rarely in
+    real usage. A situtation that triggeres this slow path is aggregations
+    where aggregations return different numbers of rows (see
+    `test_aggregation_group_by` for a specific example).
```
Comment (icexelloss) [inline ibis/backends/dask/execution/util.py:80]: Can you be more specific what the case here is? I took a look at https://github.com/ibis-project/ibis/blob/17628efdf12a2ff2f4192c4ce04dff5a86608483/ibis/backends/dask/tests/execution/test_operations.py#L118 but didn't know case do mean.

File: ibis/backends/dask/tests/execution/conftest.py [hunk 15]
```diff
@@ -71,7 +71,7 @@ def df():
             'map_of_complex_values': [None, {'a': [1, 2, 3], 'b': []}, {}],
         }
     )
-    return dd.from_pandas(pandas_df, npartitions=1)
+    return dd.from_pandas(pandas_df, npartitions=2)
```
Comment (icexelloss) [inline ibis/backends/dask/tests/execution/conftest.py:74]: Probably better to define a variable for default num partitions in case we want to change this again.

File: ibis/backends/dask/tests/execution/test_operations.py [hunk 16]
```diff
@@ -128,15 +124,16 @@ def test_aggregation_group_by(t, df, where, ibis_func, dask_func):
     )
     result = expr.execute()
 
-    dask_where = where(df)
+    dask_where = where(df.compute())
```
Comment (icexelloss) [inline ibis/backends/dask/tests/execution/test_operations.py:127]: why this change?

File: ibis/backends/dask/execution/reductions.py (hunk 8 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/reductions.py:92]: I'm not sure I understand the question?

We hit this operation when you're performing a reduction operation on a grouped series and your mask is also grouped. The pandas backend has a similar helper https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/generic.py#L522 that works a bit differently.

File: ibis/backends/tests/conftest.py [hunk 17]
```diff
@@ -192,6 +192,13 @@ def df(alltypes):
     return alltypes.execute()
 
 
+@pytest.fixture(scope='session')
+def pandas_df(backend, alltypes):
```
Comment (icexelloss) [inline ibis/backends/tests/conftest.py:196]: maybe call this `pandsa_alltypes`?

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/generic.py:155]: > How is groupby handled here?

I don't understand the question - this operation is only registered for regular `dd.Series`. 

> Also, can you create an Jira to track this issue?

Which issue?

File: ibis/backends/tests/conftest.py [hunk 18]
```diff
@@ -192,6 +192,13 @@ def df(alltypes):
...
+    if backend.name() == "dask":
+        return alltypes.execute().compute()
```
Comment (icexelloss) [inline ibis/backends/tests/conftest.py:198]: This is a bit weird. I think the interface of "execute" defines that it returns a pd.DataFrame/pd.Series. I feel it's not so great to change the interface here. Perhaps we should define a method on Dask backend to return a dd.DataFrame that is not called execute().

For example, PySpark backend seperates these two methods:
https://github.com/ibis-project/ibis/blob/master/ibis/backends/pyspark/client.py#L25

Can be a follow up

File: ibis/backends/dask/execution/reductions.py (hunk 10 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/reductions.py:85]: The pandas version has different things passed to it. It is called via `data, functools.partial(_filtered_reduction, mask.obj, method)` where `mask.obj` gives you the underlying ungrouped Series. This partial then gets applied in aggctx. In the dask version, we regroup so we can call `.sum` (or whatever) directly. They perform the same functionality, just through a different method.

**Review by icexelloss — CHANGES_REQUESTED** [review]
Reviewed one around. Left some comments.

File: ibis/backends/dask/tests/execution/test_operations.py (hunk 16 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/execution/test_operations.py:127]: Because the where clause will not work directly against a dask frame

File: ibis/backends/tests/conftest.py (hunk 17 above)
Comment (gerrymanoim) [reply ibis/backends/tests/conftest.py:196]: I called it `pandas_df` since the regular fixture called  `df` across all the test code.

File: ibis/backends/dask/execution/aggregations.py (hunk 7 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/aggregations.py:122]: I see. So the aggcontext here is a pandas aggcontext? (Would be great if you can clarify this in the code)

I wonder how did it work if we are passing a dask.Series/dask.SeriesGroupBy to the pandas aggcontext...

Comment (icexelloss) [reply ibis/backends/dask/execution/aggregations.py:122]: If the branch isn't hit/tested, perhaps put a comment I don't need to spend time reviewing it :)

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: Oh sorry I mean the "TODO" here

Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: > I don't understand the question - this operation is only registered for regular dd.Series.

Oh I see.  Where is `table.groupby('key').first()` and `table.groupby('key').first()` handled?

File: ibis/backends/tests/conftest.py (hunk 18 above)
Comment (gerrymanoim) [reply ibis/backends/tests/conftest.py:198]: The dask backend never returns a `pd.DataFrame` or `pd.Series` on execute (nor do I think it should). It always gives you back a dask task graph you are required to run `compute` on. This creates a problem for a few tests, particularly in `test_aggregations` that expect the expected test case to be a pandas `DataFrame` and support all the pandas functions.

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: (It's fine if it's not, just want to make sure I didn't miss sth)

File: ibis/backends/dask/execution/util.py (hunk 14 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:80]: Any of the cases hit this issue. The problem is that your have an aggregation that returns different numbers of rows depend on whether it is masked or not. 

https://github.com/ibis-project/ibis/blob/17628efdf12a2ff2f4192c4ce04dff5a86608483/ibis/backends/dask/tests/execution/test_operations.py#L126-L127

File: ibis/backends/dask/execution/util.py (hunk 12 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:55]: The computation?

File: ibis/backends/dask/execution/util.py (hunk 14 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/util.py:80]: Sorry I think I am missing sth here. I thought the aggregation returns n rows where n is the number of groups?

File: ibis/backends/dask/execution/reductions.py (hunk 8 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/reductions.py:92]: Aha I see. Unfortunately the pandas backend  code isn't well documented either :(

For my benefit, can you please explain what the mask is used for? Is it for the case where you have a where clause with the aggregation?

File: ibis/backends/dask/execution/reductions.py (hunk 10 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/reductions.py:85]: I see. I am curious it's reason that we don't follow pandas structure to pass a dd.SeriesGroupBy here?

File: ibis/backends/dask/execution/util.py (hunk 13 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:66]: Not entirely sure what you mean - this function itself does not enforce any sort of partitioning.

File: ibis/backends/dask/execution/aggregations.py (hunk 7 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/aggregations.py:122]: Yes - the dask backend does not currently implement its own aggegation context, but that may change when we get to more complicated operations.

File: ibis/backends/dask/execution/util.py (hunk 14 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/util.py:80]: `neg_mean_int64_with_zeros=(-t.int64_with_zeros).mean(where=ibis_where),` will only return rows (and perform the calculation) where the mask is true. 

`nunique_dup_ints` will run for each group. 

So if you're masking by your grouping column those two operations will return different numbers of rows.

File: ibis/backends/dask/execution/reductions.py (hunk 8 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/reductions.py:92]: I think https://github.com/ibis-project/ibis/pull/2596#discussion_r566222288 is the same question? Let me know if that still isn't clear.

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/generic.py:155]: Huh - for some reason that's in `numeric` via `execute_arbitrary_series_groupby`. 

Not sure why I put that there originally - I'll move it.

File: ibis/backends/dask/execution/reductions.py (hunk 10 above)
Comment (gerrymanoim) [reply ibis/backends/dask/execution/reductions.py:85]: So there are few things at work in the pandas backend. Looking that the following code

```python
def _filtered_reduction(mask, method, data):
    return method(data[mask[data.index]])


@execute_node.register(ops.Reduction, SeriesGroupBy, SeriesGroupBy)
def execute_reduction_series_gb_mask(
    op, data, mask, aggcontext=None, **kwargs
):
    method = operator.methodcaller(type(op).__name__.lower())
    return aggcontext.agg(
        data, functools.partial(_filtered_reduction, mask.obj, method)
    )
``` 
via https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/generic.py#L526

Here `aggcontext.agg` will get passed a partial that will perform the filtering and then apply the reduction method. This will get called in https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/aggcontext.py#L353, specifically via the callable branch. 

Dask does not play well with arbitrary user defined aggregations. You have to use https://docs.dask.org/en/latest/dataframe-api.html#custom-aggregation to create your own wrapper. If you pass it a partial it will not know how to work with that. Luckily dask already defines aggregations for all the common operations we have in `type(op).__name__.lower()`. So we construct the proper grouped and filtered Series and dispatch directly to dask for the actual aggregation via https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/aggcontext.py#L355

File: ibis/backends/dask/execution/generic.py (hunk 9 above)
Comment (icexelloss) [reply ibis/backends/dask/execution/generic.py:155]: Aha ok.

**Review by icexelloss — APPROVED** [review]

**Comment by icexelloss** [comment]
LGTM

**Comment by jreback** [comment]
thanks @gerrymanoim
